Add Digital Asset Manager example to app-examples

addisonschultz opened the pull request:
This PR adds an open sourced and cleaned up version of a Digital Asset Manager built with Bynder's public API. This is part of a larger project that also includes a guide on building and using this example.

mettin commented on examples/digital-asset-manager/netlify/functions/collectionId.js:
```diff
@@ -0,0 +1,36 @@
+  const cookies = parse(event.headers.cookie);
+  const access_token = cookies.bynder_access_token;
+
+  return await fetch(
```
Can we rewrite this block with async/await?

mettin commented on examples/digital-asset-manager/src/components/Assets.tsx:
```diff
@@ -0,0 +1,95 @@
+  const [assetUrls, setAssetUrls] = useState<ImageData[]>([]);
+  const [assetCount, setAssetCount] = useState(0);
+
+  const listFiles = () => {
```
This function also fetches the files from a netlify function so i don't feel like it's name captures all the responsibilities

mettin commented on examples/digital-asset-manager/src/components/CatalogItem.tsx:
```diff
@@ -0,0 +1,21 @@
+  property_extension: string;
+}) => {
+  return (
+    <div id={"catalog-item-container"}>
```
Use classnames and remove `{` + `}`

mettin commented on examples/digital-asset-manager/src/components/CategoryRow.tsx:
```diff
@@ -0,0 +1,27 @@
+  return (
+    <Link
+      to={`${slug}/${collectionId}`}
+      id={"category-row-container"}
```
Remove `{` + `}`

mettin commented on examples/digital-asset-manager/src/components/Home.tsx:
```diff
@@ -0,0 +1,45 @@
+      {signedIn ? (
+        <Categories />
+      ) : (
+        <div id={"sign-in-container"}>
```
Basically everytime you use ID, switch it to classnames, and remove { where its not needed

mettin commented on examples/digital-asset-manager/src/components/Input.tsx:
```diff
@@ -0,0 +1,63 @@
+  );
+};
+
+const X = ({ onClick }: { onClick: () => void }) => {
```
Rename to `CloseButton` and use `<button>` as the HTML tag

mettin reviewed (changes requested):
it looking ok. Just try using async await, remove id's and extra `{` / `}` characters

addisonschultz commented on examples/digital-asset-manager/netlify/functions/collections.js:
```diff
@@ -0,0 +1,35 @@
+  const cookies = parse(event.headers.cookie);
+  const access_token = cookies.bynder_access_token;
+
+  return await fetch(
```
Yep, you're right 👍

addisonschultz commented:
Thanks for the quick review @mettin 😄 

I've updated the `ids` to `classNames`, removed brackets. 

Regarding your comments about rewriting the functions to async/await blocks. Can you elaborate a bit more on this? Aren't these functions already `async`?

addisonschultz commented on examples/digital-asset-manager/src/components/BynderSignIn.tsx:
```diff
@@ -0,0 +1,41 @@
+  setSignedIn: (value: boolean) => void;
+}) => {
+  const handleSignIn = async () => {
+    const timer = setInterval(() => {
```
When clicking this button, it sends a request to the `signin`/`redirect` functions, which set the browser cookies with the access tokens needed to use the app.

I set an interval to check for the cookies every second, and if the cookies are present, the app starts to load correctly

The logic for checking the cookies is in `netlify/functions/authorize.js`

addisonschultz replied:
I'm not going to include this feature in this PR in light of time - but I've added it to the backlog for our internal version + this example. 

I have scheduled next month to work more on this!

addisonschultz commented on examples/digital-asset-manager/netlify/functions/redirect.js:
```diff
@@ -0,0 +1,76 @@
+        body: "<script>window.close();</script>",
+      };
+    })
+    .catch((error) => console.error(error));
```
There isn't because this function isn't called from the client, it's called from the redirect from Bynder. This is a bit related to the question you've asked below about the 1 second interval. 

Maybe, if there is an error, I could set a cookie called error (or something similar), and check for this in the `authorize` endpoint (And also show a visual error as needed)?

addisonschultz commented:
@mettin I've made all the requested updates 😄

mettin commented on examples/github-appcards/.netlify/functions-serve/issues/src/src/utils/github.ts:
```diff
@@ -0,0 +1,270 @@
+  )
+    .then((response) => response.json())
+    .then((result) => {
+      console.log(result);
```
remove log

addisonschultz commented on examples/digital-asset-manager/netlify/functions/authorize.js:
```diff
@@ -0,0 +1,107 @@
+  }
+
+  // Default user to signed out
+  console.log("No cookies in request");
```
Removed logs and made sure comments in code were there

addisonschultz commented on examples/digital-asset-manager/src/components/Assets.tsx:
```diff
@@ -0,0 +1,94 @@
+  name: string;
+  thumbnails: {
+    mini: string;
+    thul: string;
```
This is what is returned from Bynder API

mettin commented on examples/digital-asset-manager/src/components/Assets.tsx:
```diff
@@ -24,70 +24,72 @@ const Assets = ({
   };
 
   useEffect(() => {
     fetchAssetsByCollectionId();
```
Setting the results into state is a side effect. Return the results and put them into state from within `useEffect`

mettin replied:
You called your function `fetchAssetsByCollectionId`. As a reader, I'd expect that to return something, not change anything in state. So a solution would be to rename the function to `fetchAssetsByCollectionIdAndSetThemInState`, which is a bit of a mess. Separation of concerns would be better here:


```ts
useEffect(() => {
  const getData = async () => {
  	const assets = await fetchAssetsByCollectionId();
    setThemInstate(assets);
  }
 
  getData();
  
})
```

mettin commented on examples/digital-asset-manager/src/components/Assets.tsx:
```diff
@@ -24,70 +24,72 @@ const Assets = ({
+        <>
+          <div className="asset-item-container">
+            {assetUrls
+              .filter((assetObject) => {
```
Does it make sense to do this check inside the `.map`? Just return `null` when its not needed. Will result in 1 less loop over the array elements. Only needed if this is a BIG array imho. Microwin

addisonschultz replied:
Wouldn't a filter make more sense when you need to filter something out? 😄

mettin replied:
yeah it does, only when you have A lot of items in the array (say + 5k) this will prevent a loop of 5k items. So i dont think we need it here

mettin commented on examples/digital-asset-manager/src/components/Assets.tsx:
```diff
@@ -24,70 +24,72 @@ const Assets = ({
+                      <img
+                        src={asset.thumbnails.thul}
+                        data-image={asset.thumbnails.webimage}
+                        className="miro-draggable draggable-item asset-image-background"
```
no alt?

mettin commented on examples/digital-asset-manager/src/components/Categories.tsx:
```diff
@@ -16,24 +16,22 @@ const Categories = () => {
-    });
+  const listBynderCollections = async () => {
+    try {
+      await fetch(`/.netlify/functions/authorize`);
```
You're not storing the result of this, dont you need to check if the authorization went well?

addisonschultz replied:
For this call no, this authorize call will only be used to refresh the token pair - it will refresh the tokens if they are expired before calling the collections endpoint

mettin reviewed (changes requested):
Using async/await is just more readable IMHO. I don't see promises.then notation being used anymore that often. Of you feel like this is good enough then ignore my comment

mettin commented on examples/digital-asset-manager/src/components/Assets.tsx:
```diff
@@ -77,6 +81,7 @@ const Assets = ({
                         data-image={asset.thumbnails.webimage}
                         className="miro-draggable draggable-item asset-image-background"
                         draggable={false}
+                        alt="Draggable Image"
```
This doesnt tell anything about the images contents
